File: recursion/backtracking/CombinationSum.cpp
class CombinationSum {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> combSet;
        if (candidates.size() == 0) {
            return combSet;
        }
        vector<int> combination;
        getCombination(candidates, target, 0, combSet, combination);
        return combSet;
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> combSet;
        if (candidates.size() == 0) {
            return combSet;
        }
        sort(candidates.begin(), candidates.end());
        vector<int> combination;
        noReuseCombination(candidates, target, 0, combSet, combination);
        return combSet;
    }

private:
    void getCombination(vector<int>& candidates, int target, int startPos,
        vector<vector<int>>& combSet, vector<int>& combination) {
        if (target == 0) {
            combSet.push_back(combination);
            return;
        }
        if (target < 0) {
            return;
        }
        for (int i = startPos; i < candidates.size(); i++) {
            combination.push_back(candidates[i]);
            getCombination(candidates, target - candidates[i], i, combSet, combination);
            combination.pop_back();
        }
    }

    void noReuseCombination(vector<int>& candidates, int target, int startPos,
        vector<vector<int>>& combSet, vector<int>& combination) {
        if (target == 0) {
            combSet.push_back(combination);
            return;
        }
        if (target < 0) {
            return;
        }
        for (int i = startPos; i < candidates.size(); i++) {
            if (i != startPos && candidates[i] == candidates[i - 1]) {
                continue;
            }
            combination.push_back(candidates[i]);
            noReuseCombination(candidates, target - candidates[i], i + 1, combSet, combination);
            combination.pop_back();
        }
    }
};

File: recursion/backtracking/CombinationSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CombinationSum {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> combSets = new ArrayList<>();
        if (candidates == null || candidates.length == 0) {
            return combSets;
        }
        // Arrays.sort(candidates);
        combinationSum(candidates, target, 0, combSets, new ArrayList<>());
        return combSets;
    }

    private void combinationSum(int[] candidates, int target,
            int candidatePos, List<List<Integer>> combSets, List<Integer> oneSum) {
        if (target == 0) {
            combSets.add(new ArrayList<>(oneSum));
            return;
        }
        if (target < 0) { // not equal to the desired sum, do nothing
            return;
        }
        // i starts with the initial position of the last unexhausted candidate
        for (int i = candidatePos; i < candidates.length; i++) {
            oneSum.add(candidates[i]);
            // ensure that we can reuse candidates[i] later -- but not duplicate ones
            // i.e. adding candidate[i] to the oneSum twice or more
            combinationSum(candidates, target - candidates[i], i, combSets, oneSum);
            oneSum.remove(oneSum.size() - 1);
        }
    }

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<List<Integer>> combSets = new ArrayList<>();
        if (candidates == null || candidates.length == 0) {
            return combSets;
        }
        Arrays.sort(candidates);
        combinationSum2(candidates, target, 0, combSets, new ArrayList<>());
        return combSets;
    }

    private void combinationSum2(int[] candidates, int target, int candidatePos,
            List<List<Integer>> combSets, List<Integer> oneSum) {
        if (target == 0) {
            combSets.add(new ArrayList<>(oneSum));
            return;
        }
        if (target < 0) {
            return;
        }
        for (int i = candidatePos; i < candidates.length; i++) {
            // avoid duplicate set invoked by using identital elements
            // would result in duplcate combinations added to the set otherwise
            if (i != candidatePos && candidates[i] == candidates[i - 1]) {
                continue;
            }
            oneSum.add(candidates[i]); // update candidate position: avoid reusing the same element
            combinationSum2(candidates, target - candidates[i], i + 1, combSets, oneSum);
            oneSum.remove(oneSum.size() - 1);
        }
    }
}
